### PR TITLE
Add positivity checker

### PR DESCRIPTION
**Pull request by nikivazou**

### CODE REVIEW

File: src/Language/Haskell/Liquid/GHC/Play.hs [hunk 1]
```diff
@@ -10,12 +10,98 @@ import Prelude hiding (error)
...
+-------------------------------------------------------------------------------
+
+getNonPositivesTyCon :: [TyCon] -> [(TyCon, [DataCon])]
+getNonPositivesTyCon tcs = Mb.catMaybes $ map go (M.toList $ makeAppearences tcs)
+  where 
+    go (tc,dcocs) = case filter (\(_,occ) -> elem tc (negA occ)) dcocs of  
+                      [] -> Nothing
+                      xs -> Just (tc, fst <$> xs) 
+
+data TyConAppearance = TyConA {posA :: [TyCon], negA :: [TyCon] }
```
Comment (facundominguez) [inline src/Language/Haskell/Liquid/GHC/Play.hs]: How about documenting the meaning of the fields in this type?

File: src/Language/Haskell/Liquid/GHC/Play.hs [hunk 2]
```diff
@@ -10,12 +10,98 @@ import Prelude hiding (error)
...
+import Language.Haskell.Liquid.GHC.API as Ghc hiding (substTysWith, panic,showPpr)
 import Language.Haskell.Liquid.GHC.Misc ()
 import Language.Haskell.Liquid.Types.Errors
 
 
+-------------------------------------------------------------------------------
+-- | Positivity Checker -------------------------------------------------------
+-------------------------------------------------------------------------------
+
+getNonPositivesTyCon :: [TyCon] -> [(TyCon, [DataCon])]
```
Comment (facundominguez) [inline src/Language/Haskell/Liquid/GHC/Play.hs:28]: How about documenting the output of this function? Perhaps documenting an example of an input and an output would be telling.

File: src/Language/Haskell/Liquid/Types/Errors.hs [hunk 3]
```diff
@@ -1043,6 +1049,10 @@ ppError' _ dCtx (ErrRewrite _ msg )
         $+$ dCtx
         $+$ nest 4 msg
 
+ppError' _ dCtx (ErrPosTyCon _ tc dc)
+  = text "Non strictly positive occurence of" <+> tc <+> "in" <+> dc 
+        $+$ dCtx
```
Comment (facundominguez) [inline src/Language/Haskell/Liquid/Types/Errors.hs:1054]: If the error could reference a github issue or some other pointer discussing the purpose of the check the user might have a better time figuring it out.

Also, maybe the error message could mention the flag to disable the check and warn appropriately about using it.

**Review by facundominguez — COMMENTED** [review]
I didn't try to understand the code, but made a pass trying to identify what could be interesting to have a bit more explained.

**Comment by ranjitjhala** [comment]
These are excellent suggestions @facundominguez !

**Comment by nikivazou** [comment]
Indeed, great suggestions, esp. the documenting what can go wrong with the non positive occurrences! 
I will do it tomorrow. 

I was happily surprised that only one test failed! (All these complaints for only one test :))

**Comment by ranjitjhala** [comment]
Complaints? You are referring to the oopsla reviewers right? :-)

**Comment by nikivazou** [comment]
Not only. 

This check addresses #159 and the known TT fact that negative occurrences are  bad (https://stackoverflow.com/questions/12651146/why-inductive-datatypes-forbid-types-like-data-bad-a-c-bad-a-a-where-th).

**Comment by facundominguez** [comment]
The comments look great, thank you!

**Pull request merged**